GitHubCodespaces.ps1

```diff
@@ -0,0 +1,945 @@
+            }
+
+            break
+        }
+
+        { $_ -in ('Elements', 'Uri') }
+        {
+            $elements = Resolve-RepositoryElements
+            $OwnerName = $elements.ownerName
+            $RepositoryName = $elements.repositoryName
+
+            $telemetryProperties['OwnerName'] = Get-PiiSafeString -PlainText $OwnerName
+            $telemetryProperties['RepositoryName'] = Get-PiiSafeString -PlainText $RepositoryName
+
+            $uriFragment = "repos/$OwnerName/$RepositoryName/codespaces"
+            $description = "Getting $OwnerName/$RepositoryName/codespaces"
+
+            break
+        }
+    }
+
+    $params = @{
+        UriFragment = $uriFragment
+        Description = $description
+        AccessToken = $AccessToken
+        TelemetryEventName = $MyInvocation.MyCommand.Name
+        TelemetryProperties = $telemetryProperties
+    }
+
+    $result = Invoke-GHRestMethodMultipleResult @params
+    if ($null -ne $result.codespaces)
+    {
+        $result = $result.codespaces
+    }
+
+    return ($result | Add-GitHubCodespaceAdditionalProperties)
+}
+
+function New-GitHubCodespace
+{
+    <#
+    .SYNOPSIS
+        Creates a codespace.
+
+    .DESCRIPTION
+        Creates a codespace.
+
+        The Git repo for this module can be found here: http://aka.ms/PowerShellForGitHub
+
+    .PARAMETER OwnerName
+        Owner of the Codespace.
+        If not supplied here, the DefaultOwnerName configuration property value will be used.
+
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+
+    .PARAMETER Uri
+        Uri for the Codespace.
+        The OwnerName and CodespaceName will be extracted from here instead of needing to provide
+        them individually.
+
+    .PARAMETER PullRequest
+        The pull request number for this codespace.
+
+    .PARAMETER RepositoryId
+        The ID for a Repository.  Only applicable when creating a codespace for the current authenticated user.
+
+    .PARAMETER Ref
+        Git ref (typically a branch name) for this codespace
+
+    .PARAMETER ClientIp
+        IP for location auto-detection when proxying a request.
+
+    .PARAMETER Devcontainer
+        Path to devcontainer.json config to use for this codespace.
+
+    .PARAMETER DisplayName
+        Display name for this codespace
+
+    .PARAMETER Location
+        The requested location for a new codespace.
+        Best efforts are made to respect this upon creation.
+        Assigned by IP if not provided.
+
+    .PARAMETER Machine
+        Machine type to use for this codespace.
+
+    .PARAMETER NoMultipleRepoPermissions
+        Whether to authorize requested permissions to other repos from devcontainer.json.
+
+    .PARAMETER RetentionPeriod
+        Duration in minutes (up to 30 days) after codespace has gone idle in which it will be deleted.
+
+    .PARAMETER Timeout
```

Comment:
I think both `Timeout` and `RetentionPeriod` should indicate their unit of measurement in their name (`TimeoutMinutes`, `IdleRetentionPeriodMinutes`).  Would help users not to have to read the documentation to understand what the time unit is.  I also think we should keep `Idle` in the retention period name.  Users can still just use `Timeout` and `IdleRetentionPeriod` when specifying the parameters since PowerShell doesn't require full parameter names when there's no ambiguity, but I think the default for tab completion should be with `Minutes` so that users know.

Reply:
as you wish...